Add CircleCI

ruffsl opened the pull request:
This PR add CI support for CircleCI.

# Summary

Like Travis, CircleCI is a Continuous Integration service, however CircleCI provides more docker integration such as support for custom images without resorting to machine type executors. This means that CI jobs can start as soon as they are triggered, rather than queuing for an available Virtual Machine to then spin up. The pulling of the Docker Image is perhaps the only main flat overhead in triggering jobs. Further features such as caching and storage of artifacts additionally accelerate job times while providing insight into build and test results.

---

# Overview

The proposed CI would start with pre-build Docker Image, hosted on automated repo on Docker Hub that would trigger tag builds upon new commits into particular GitHub branches or rebuilds of linked Docker Hub repos, i.e. `osrf/ros2:nightly` base image. Built tags would include all/most the necessary dependencies for the project, include any build dependencies for prerequisite dependencies, i.e. running `vcs import`, `rosdep install`, etc. on the master branch. Even when short cutting the install time by building `FROM` the `osrf/ros2:nightly` base image, this takes about 90 min on Docker Hub, so it helpful to offload/bake this amount of setup into the image our CI pulls from.

Once the CI pulls the docker image above, the config proceeds to clean the workspaces and checkout the source to inspect. This also entails `vcs import`ing and prerequisite defined in the new checked out source, installing the prerequisite dependencies, building the prerequisites themselves, installing the source packages dependencies, building the source packages, then testing the source packages.

In between steps in the job of the workflow, caching of the build stages for both prerequisite and source packages is attempted. Caches are broken with either the installed dependencies change or the source code has changed. This allows the cache to persist across jobs.

> An example of current table of caching in action:

| prerequisites | source build | source tests | time |
|---|---|---|---|
| w/o | w/o | w/o | [16:19](https://circleci.com/gh/ruffsl/navigation2/58) |
| w  | w/o | w/o | [05:07](https://circleci.com/gh/ruffsl/navigation2/62)
| w   | w  | w/o | [02:07](https://circleci.com/gh/ruffsl/navigation2/60) |

Lastly, the log files for the colcon build and test steps are also archived and made linkable by storing them with the job's public artifacts.

![image](https://user-images.githubusercontent.com/2293573/51814617-1033f480-2271-11e9-8cf7-8a8e8e6154a2.png)


## Workflow

- DockerHub builds CI Image
  - Triggered either from linked base image or branch commits
  - Shortcuts CI by installing the majority of decencies
- CircleCI
  - Triggered by any workflow condition, e.g. commit, PR, web-hook
  - Pulls CI image from DockerHub
  - Cleans source and prerequisite workspaces
  - Checks out respective source from trigger
  - Prerequisite
    - Checks out prerequisite from .repo file in source
    - Installs any new/missing prerequisite dependencies
    - Restores prerequisite build cache if possible
    - Builds prerequisite if cache restore falls through
    - Saves prerequisite build cache if needed
  - Source
    - Installs any new/missing source dependencies
    - Restores source build cache if possible
    - Builds source if cache restore falls through
    - Saves source build cache if needed
    - Save build log artifacts
  - Test
    - Test source build
    - Save test log artifacts

# Action Items

- DockerHub
  - [x] Create the `ros-planning` DockerHub organization (admin)
  - [x] Invite relevant Collaborators to DockerHub org (admin)
  - [x] Link GitHub `navigation2` automated repo to DockerHub org (admin)
  - [x] Configure build rules for automated repo (@ruffs)
    - [x] Add rule for branchs, e.g. `ros-planning/navigation2:master`
    - [x] Add repo link for Base Image: `osrf/ros2:nightly`

- CircleCI
  - [x] Enable GitHub integration (admin)
  - [x] Invite relevant Collaborators to GitHub Team (admin)
  - [x] Configure Advanced Settings for ci repo (@ruffs)
    - [x] Only build pull requests (**On**/Off)?
      - Save build usage by only testing open PRed branches
    - [x] Auto-cancel redundant builds (**On**/Off)?
      - Save build usage by only testing current HEAD of branch
    - [x] Build forked pull requests (**On**/Off)?
      - Looking at repo's history, most merged PRs seem to come from forks

---

# References:

## CircleCI
* [Configuration Reference](https://circleci.com/docs/2.0/configuration-reference/)
  * To get the gist of the config file in PR
* [Building Open Source Projects](https://circleci.com/docs/2.0/oss/)
  * Special considerations for Open Source Projects
* [Migrating From Travis CI](https://circleci.com/docs/2.0/migrating-from-travis/)
  * Other notes/suggestions and highlighting differences in CI
* [Caching](https://circleci.com/docs/2.0/caching/)
  * More on Caching Strategy Tradeoffs when used between workflows and jobs
* [Debugging with SSH](https://circleci.com/docs/2.0/ssh-access-jobs/)
  * This was helpful when writing/debugging this config
* [Caching apt-get packages](https://discuss.circleci.com/t/caching-apt-get-packages/9076)
  * Using 2.0, this is now done by priming the docker image
* [Storing Build Artifacts](https://circleci.com/docs/2.0/artifacts/)
  * Used make colcon build and test log files more accessible

## Related:
 * https://github.com/ros-planning/navigation2/pull/525
 * https://github.com/osrf/docker_images/pull/202
 * https://github.com/osrf/docker_images/issues/201

crdelsey commented:
Is the performance gain largely because you've switched to CircleCI or is it because you're starting from `ruffsl/navigation2:master`? If we switch to CircleCI, do we still need to have Travis as well, or can we do everything within Circle to keep things simpler?

ruffsl commented:
> Is the performance gain largely because you've switched to CircleCI or is it because you're starting from ruffsl/navigation2:master?

@crdelsey , its partly both. Most of the upfront time is saved by starting from a Docker Image prepped for the CI with common dependence pre-installed, but then some more is also saved by making use of the caching between jobs. We could perhaps save even more by enabling [Docker Layer Caching](https://circleci.com/docs/2.0/docker-layer-caching/), but this currently seems to be a premium feature from CircleCI.

> If we switch to CircleCI, do we still need to have Travis as well, or can we do everything within Circle to keep things simpler?

Well, I'm not sure. You could have both if you wanted, don't think it would be much drain. You might want to have only one or the other CI enabled for PRs to prevent too much integration noise. CircleCI might be better for this, as long as we can keep the quick turnaround times I've tabled above.

I'm not sure Travis has the same [monthly limit](https://circleci.com/pricing) as CircleCI, so it may still be well suited for longer builds via periodic cron jobs. E.g. just rebuilding the whole Dockerfile daily via master branch and run the stock tests and code coverage.

Also ping @mkhansen-intel , @SteveMacenski  , I've updated the top PR comment with more details.

ruffsl commented on .circleci/config.yml:
```diff
@@ -0,0 +1,145 @@
+            tr -d \, | xargs -n1 | sort -u | xargs)
+          dpkg -s $dependencies | \
+            sha256sum | \
+            awk '{print "dependencies "$1}' >> $ROS_WS/checksum.txt
```
Looking at this again, I'm not sure this alone would be a sufficient cache break. It may likely be that core ros packages installed by the nightly flat archives from the base image could change; and as they aren't debs, listing from dpkg would not generate a diff. Perhaps we could append a digest of `/opt/ros/$ROS_DISTRO/` to `$ROS_WS/checksum.txt`?

This would also essentially break the cache after every new CI docker image update, but I don't spot an immediate workaround. How often are the prerequisites in the `ros2_dependencies.repos` changed in PRs? We could comment out the prerequisites steps in the build job, and let the CI docker image keep them uptodate; and if someone does need to test a change there, they only need to uncomment lines L119 and L121.

ruffsl replied:
Resolved with https://github.com/ros-planning/navigation2/pull/534#issuecomment-459842619

crdelsey commented:
docker hub doesn't allow symbols in an organization name, so I called the org `rosplanning` instead of `ros-planning`

ruffsl commented:
Ok, I've polished up the caching chain so it should skip the upstream re-build if the `.repo` file is unchanged from what already existed in the CI docker image, as well as breaking the cache if the ros distro files update. Also updated the test the ctest retry to match travis, and updated the image used to point to the `rosplanning` org.

> docker hub doesn't allow symbols in an organization name, so I called the org rosplanning instead of ros-planning

Sounds fine. @crdelsey , it looks like these items were left undone, so my own account can not complete the first due to the absence of the second.

>  - [ ] Link GitHub navigation2 automated repo to DockerHub org (admin)
>  - [ ] Invite relevant Collaborators to GitHub Team (admin)

I.e I'm unable to setup the Build configurations for the `rosplanning/navigation2`


![image](https://user-images.githubusercontent.com/2293573/52144502-54c5e400-2613-11e9-8d71-5110fbd50d2b.png)


![image](https://user-images.githubusercontent.com/2293573/52144818-301e3c00-2614-11e9-9eb2-72b2907f7f28.png)

crdelsey commented:
I just linked DockerHub to Github, so you should have access now. I also set up the Dockerhub build as you showed above and triggered it.

ruffsl commented:
Alright, 2f02de1 is passing on my end here:  
https://circleci.com/gh/ruffsl/navigation2/80

I don't see the CircleCI from ros-planning though:  
https://circleci.com/gh/ros-planning/navigation2

@crdelsey, did you enable `Build forked pull requests` in the CircleCI repo settings for navigation2?
![image](https://user-images.githubusercontent.com/2293573/52153045-64eabd00-262d-11e9-892e-11163c058fa0.png)

After we see green light from ros-planning's circleci on this PR, this should be good to merge.
I may have to rebase or something to retrigger a fork PR job after that setting is enabled.

crdelsey commented:
@ruffsl I did enable "Build forked pull requests". I made you an admin of this repo so I'm not slowing things down. You should be able to check any settings directly.

crdelsey commented:
@ruffsl Circle CI is running on #553, so it could just be that it doesn't run on this PR because it was created before CircleCI was enabled. I'm willing to just merge to master and see what happens. We can then quickly fix any problems that come up or just revert the change if the problem is big, which seems unlikely.

crdelsey commented:
@orduno @mhpanah @mkhansen-intel Can one of you review this so we can merge?

ruffsl commented:
Well, after the merge commit triggered a fresh CI job, it looks like the CTest for `nav2_dynamic_params` is a little flaky. Remember that the test logs are accessible via the artifacts tab:

https://circleci.com/gh/ros-planning/navigation2/16#artifacts

Although the artifacts tab may seem invisible if your not logging into github, one can force it visible by just adding `#artifacts` to the job url. Related Bug Report:

https://discuss.circleci.com/t/failing-to-upload-artifacts-from-symbolic-link-directories/28000

crdelsey commented:
I opened #556 for the flaky test. By the way, thanks for doing all this work cleaning up the CI stuff for us. I really appreciate it!
